### PR TITLE
[go_router] feat: allow pass custom builder functions to `GoRouter` class

### PR DESCRIPTION
**Pull request by leynier**
For instantiate the following properties:

- `backButtonDispatcher`
- `configuration`
- `routeInformationParser`
- `routeInformationProvider`
- `routerDelegate`

By default use the actual constructors

Part of https://github.com/flutter/flutter/issues/119761

In this [comment](https://github.com/flutter/flutter/issues/119761#issuecomment-1637258172), I explain the changes proposed by this pull request.

Resolve https://github.com/flutter/flutter/issues/119761

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [ ] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/go_router/lib/src/router.dart [hunk 1]
```diff
@@ -24,6 +25,41 @@ typedef GoExceptionHandler = void Function(
...
+
+/// The function signature of [RouteConfiguration.new]
+typedef RouteConfigurationBuilder = RouteConfiguration Function({
+  required List<RouteBase> routes,
+  required int redirectLimit,
+  required GoRouterRedirect topRedirect,
+  required GlobalKey<NavigatorState> navigatorKey,
+});
+
+/// The function signature of [GoRouteInformationParser.new]
```
[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/go_router/lib/src/router.dart:39]: same for other

**Review by chunhtai — COMMENTED** [review]
Can you also add some tests? This also require a bump to minor version

**Comment by chunhtai** [comment]
Hi @leynier are you still planning on coming back to this pr?

**Comment by leynier** [comment]
> Hi @leynier are you still planning on coming back to this pr?

Sorry for the late response, yes, I plan to continue with the PR because I'm interested in having that functionality.

**Comment by chunhtai** [comment]
(triage) @leynier Is it ok to close this pr for now until you have time to come back to this PR?

**Comment by leynier** [comment]
> (triage) @leynier Is it ok to close this pr for now until you have time to come back to this PR?

Yes, sure, no problem 👍

**Pull request closed without merging**